### PR TITLE
RS/YJ/Rule 12-4

### PR DESCRIPTION
**Pull request by yunjoonjung-PNNL**
This rule hasn't been tested. I'll revisit it once the testing file and ```is_space_a_computer_room``` function are developed. Thanks!

### CODE REVIEW

**Comment by yunjoonjung-PNNL** [comment]
The rule was tested with Juan's PR and all the cases passed.

File: rct229/rulesets/ashrae9012019/section12/section12rule4.py [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+                mask_schedule = [1] * 8784 if is_leap_year else [1] * 8760
+
+                comparison_data = compare_schedules(
+                    multiplier_schedule_b,
+                    expected_hourly_values,
+                    mask_schedule,
+                    is_leap_year,
+                )["total_hours_matched"]
+
+                return {"comparison_data": comparison_data}
```
Comment (weilixu) [inline rct229/rulesets/ashrae9012019/section12/section12rule4.py:153]: Rename this parameter to total_hours_matched.
Also, let's add another parameter `hours_in_a_year = LeapYear.LEAP_YEAR_HOURs if is_leap_year else LeapYear.REGULAR_YEAR_HOURS`

File: rct229/rulesets/ashrae9012019/section12/section12rule4.py [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+                hourly_multiplier_schedule_b = misc_equip_b["multiplier_schedule"]
+
+                multiplier_schedule_b = schedule_b[hourly_multiplier_schedule_b]
+                expected_hourly_values = []
+                for month in range(1, 13):
+                    expected_hourly_values.extend(
+                        [MONTH_FRACTIONS[month]] * DAYS_IN_MONTH[month] * 24
+                    )
+
+                mask_schedule = [1] * 8784 if is_leap_year else [1] * 8760
```
Comment (weilixu) [inline rct229/rulesets/ashrae9012019/section12/section12rule4.py:144]: We should have the enum in the ruleset called LeapYear.LEAP_YEAR_HOUR and LeapYear.REGULAR_YEAR_HOURS

File: rct229/rulesets/ashrae9012019/section12/section12rule4.py [hunk 3]
```diff
@@ -0,0 +1,159 @@
...
+                    is_leap_year,
+                )["total_hours_matched"]
+
+                return {"comparison_data": comparison_data}
+
+            def rule_check(self, context, calc_vals=None, data=None):
+                comparison_data = calc_vals["comparison_data"]
+                is_leap_year = data["is_leap_year"]
+
+                return comparison_data == (8784 if is_leap_year else 8760)
```
Comment (weilixu) [inline rct229/rulesets/ashrae9012019/section12/section12rule4.py:159]: in here, you just need to match the `total_hours_matched == hours_in_a_year`

File: rct229/rulesets/ashrae9012019/section12/section12rule4.py (hunk 1 above)
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section12/section12rule4.py:153]: Good point! I keep forgetting we have the ```LeapYear``` class. Addressed.

File: rct229/rulesets/ashrae9012019/section12/section12rule4.py [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+            ):
+                if is_space_a_computer_room(rmd_b, space_b["id"]):
+                    for misc_equip_b in space_b["miscellaneous_equipment"]:
+                        if (
+                            misc_equip_b.get("type")
+                            != MISCELLANEOUS_EQUIPMENT.INFORMATION_TECHNOLOGY_EQUIPMENT
+                        ):
+                            return False
+                        else:
+                            return True
```
[the remark this answers is not in the document]
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section12/section12rule4.py:93]: Good point. Addressed.

File: rct229/rulesets/ashrae9012019/section12/section12rule4.py (hunk 2 above)
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section12/section12rule4.py:144]: Addressed.

File: rct229/rulesets/ashrae9012019/section12/section12rule4.py (hunk 3 above)
Comment (yunjoonjung-PNNL) [reply rct229/rulesets/ashrae9012019/section12/section12rule4.py:159]: Addressed.

**Pull request merged**